File: frontend/download_geojson.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

try {
  require.resolve('osmtogeojson');
} catch (e) {
  console.log('Installing osmtogeojson...');
  execSync('npm install osmtogeojson', { stdio: 'inherit' });
}

const osmtogeojson = require('osmtogeojson');

const CITIES = [
    'Cavite',
    'Dasmariñas',
    'Imus',
    'General Trias',
    'Bacoor',
    'Trece Martires',
    'Tagaytay',
    'Carmona',
    'Cavite City',
    'Silang',
    'Amadeo',
    'Mendez',
    'Indang',
    'Alfonso',
    'General Emilio Aguinaldo',
    'Maragondon',
    'Ternate',
    'Naic',
    'Tanza',
    'Noveleta',
    'Rosario',
    'Kawit'
];

async function downloadCity(city) {
    console.log(`Downloading ${city}...`);
    const fileName = city.replace(/ /g, '_').toLowerCase();
    const filePath = path.join(__dirname, 'public/geojson', `${fileName}.geojson`);
    if (fs.existsSync(filePath) && fs.statSync(filePath).size > 1000) {
        console.log(`Skipping ${city}`);
        return;
    }

    let query;
    if (city === 'Cavite') {
        query = `[out:json];relation["name"="Cavite"]["admin_level"="4"];out body;>;out skel qt;`;
    } else {
        query = `[out:json];area["name"="Cavite"]->.a;relation["name"="${city}"](area.a)["admin_level"~"6|8|9"];out body;>;out skel qt;`;
    }

    try {
        const res = await fetch('https://overpass-api.de/api/interpreter', {
            method: 'POST',
            body: query
        });
        const data = await res.json();
        const geojson = osmtogeojson(data);
        
        const feature = geojson.features.find(f => f.geometry.type === 'Polygon' || f.geometry.type === 'MultiPolygon');
        if (feature) {
            fs.writeFileSync(filePath, JSON.stringify({type: "FeatureCollection", features: [feature]}));
            console.log(`Saved ${city}`);
        } else {
            console.log(`No valid geometry found for ${city}`);
        }
    } catch (e) {
        console.log(`Failed for ${city}:`, e.message);
    }
    
    // sleep
    await new Promise(r => setTimeout(r, 2000));
}

async function main() {
    fs.mkdirSync(path.join(__dirname, 'public/geojson'), {recursive:true});
    for (const city of CITIES) await downloadCity(city);
}
main();

File: download_geojson.py
import os
import time
import json
import urllib.request
import urllib.parse

CITIES = [
    'Cavite',
    'Dasmariñas',
    'Imus',
    'General Trias',
    'Bacoor',
    'Trece Martires',
    'Tagaytay',
    'Carmona',
    'Cavite City',
    'Silang',
    'Amadeo',
    'Mendez',
    'Indang',
    'Alfonso',
    'General Emilio Aguinaldo',
    'Maragondon',
    'Ternate',
    'Naic',
    'Tanza',
    'Noveleta',
    'Rosario',
    'Kawit'
]

os.makedirs('frontend/public/geojson', exist_ok=True)

# Important to have a realistic User-Agent to avoid immediate blocks from Nominatim
headers = {
    'User-Agent': 'CaviteAgriWatch/1.0 (carlo.dev@example.com)',
    'Accept': 'application/json',
    'Accept-Language': 'en-US,en;q=0.5'
}

for city in CITIES:
    file_path = f"frontend/public/geojson/{city.replace(' ', '_').lower()}.geojson"
    
    if os.path.exists(file_path):
        size = os.path.getsize(file_path)
        if size > 1000:
            print(f"Skipping {city}, already exists.")
            continue
            
    q = "Cavite Philippines" if city == 'Cavite' else f"{city} Cavite Philippines"
    url = f"https://nominatim.openstreetmap.org/search?q={urllib.parse.quote(q)}&polygon_geojson=1&format=geojson"
    
    print(f"Downloading {city}...")
    req = urllib.request.Request(url, headers=headers)
    
    try:
        with urllib.request.urlopen(req) as response:
            data = response.read().decode('utf-8')
            with open(file_path, "w", encoding="utf-8") as f:
                f.write(data)
            print(f"Saved {city} successfully.")
    except urllib.error.HTTPError as e:
        print(f"Failed to download {city}: HTTP {e.code} {e.reason}")
        if e.code == 429:
            print("Rate limited. Pausing for 10 seconds before continuing...")
            time.sleep(10.0)
    except Exception as e:
        print(f"Failed to download {city}: {e}")
        
    time.sleep(3.0)  # Wait 3 seconds to respect Nominatim limits
